Spigot-Server-Patches/0622-Strip-raytracing-for-EntityLiving-hasLineOfSight.patch

```diff
@@ -0,0 +1,136 @@
++        RayTrace.BlockCollisionOption raytrace_blockcollisionoption = RayTrace.BlockCollisionOption.COLLIDER;
++        VoxelShape voxelshape = raytrace_blockcollisionoption.get(iblockdata, this, blockposition, voxelshapecoll); // Collider doesn't actually need a voxelshape
++        MovingObjectPositionBlock movingobjectpositionblock = this.rayTrace(vec3d, vec3d1, blockposition, voxelshape, iblockdata);
++
++        return movingobjectpositionblock == null ? null : movingobjectpositionblock.getType(); // can't be null because of the logic above
++    }
++    // Paper end
++
+     // CraftBukkit start - moved block handling into separate method for use by Block#rayTrace
+     default MovingObjectPositionBlock rayTraceBlock(RayTrace raytrace1, BlockPosition blockposition) {
+             // Paper start - Prevent raytrace from loading chunks
+diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
+index 9ed21f434c5fb019b74dfe9ee0b802ccc5c07fd8..f00c8a105c6008a34b0082bfef2826051f8a6451 100644
+--- a/src/main/java/net/minecraft/server/World.java
++++ b/src/main/java/net/minecraft/server/World.java
+@@ -256,6 +256,83 @@ public abstract class World implements GeneratorAccess, AutoCloseable {
+         return null;
+     }
+ 
++    // Paper start - broken down method of raytracing for EntityLiving#hasLineOfSight, replaces IBlockAccess#rayTrace(RayTrace)
++    protected MovingObjectPosition.EnumMovingObjectType rayTraceDirect(Vec3D vec3d, Vec3D vec3d1, VoxelShapeCollision voxelshapecoll) {
++        // most of this code comes from IBlockAccess#a(RayTrace, BiFunction, Function), but removes the needless functions
++        if (vec3d.equals(vec3d1)) {
++            return MovingObjectPosition.EnumMovingObjectType.MISS;
++        }
++        double d0 = MathHelper.d(-1.0E-7D, vec3d1.x, vec3d.x);
++        double d1 = MathHelper.d(-1.0E-7D, vec3d1.y, vec3d.y);
++        double d2 = MathHelper.d(-1.0E-7D, vec3d1.z, vec3d.z);
++        double d3 = MathHelper.d(-1.0E-7D, vec3d.x, vec3d1.x);
++        double d4 = MathHelper.d(-1.0E-7D, vec3d.y, vec3d1.y);
++        double d5 = MathHelper.d(-1.0E-7D, vec3d.z, vec3d1.z);
++        int i = MathHelper.floor(d3);
++        int j = MathHelper.floor(d4);
++        int k = MathHelper.floor(d5);
++        BlockPosition.MutableBlockPosition blockposition_mutableblockposition = new BlockPosition.MutableBlockPosition(i, j, k);
++        Chunk chunk = this.getChunkIfLoaded(blockposition_mutableblockposition);
++        if (chunk == null) {
++            return MovingObjectPosition.EnumMovingObjectType.MISS;
++        }
++
++        MovingObjectPosition.EnumMovingObjectType t0 = this.rayTraceBlockDirect(vec3d, vec3d1, blockposition_mutableblockposition, chunk.getType(blockposition_mutableblockposition), voxelshapecoll);
++
++        if (t0 != null) {
++            return t0;
++        } else {
```

Comment:
You already did a `return` for the `if`. This is unnecessary, ugly indentation.

Reply:
It's to maintain the structure of the original code, to make it easier to see what's stripped and compare.